### PR TITLE
Add `SpectrumDataset.stack()`

### PR DESCRIPTION
**Pull request by registerrier**
This PR adds a `stack()` method to `SpectrumDataset`. The stacking is done in-place. 

To properly deal with the `safe_mask` when it is not set, I have added a property to have it return a vector of `True` values.

### CODE REVIEW

File: gammapy/spectrum/dataset.py [hunk 1]
```diff
@@ -398,6 +409,66 @@ def create(cls, e_reco, e_true=None, reference_time="2000-01-01"):
...
+       """
+
+        if not isinstance(other, SpectrumDataset):
+            raise TypeError("Incompatible types for SpectrumDataset stacking")
+
+        if self.counts is not None:
+            self.counts.data[~self.mask_safe] *= 0
+            self.counts.data[other.mask_safe] += other.counts.data[other.mask_safe]
+        else:
+            self.counts = None
```
Comment (adonath) [inline gammapy/spectrum/dataset.py]: If counts is already `None` I think there is no need to set it to `None` again?

File: gammapy/spectrum/dataset.py [hunk 2]
```diff
@@ -398,6 +409,66 @@ def create(cls, e_reco, e_true=None, reference_time="2000-01-01"):
...
+
+        if self.background is not None:
+            self.background.data[~self.mask_safe] *= 0
+            self.background.data[other.mask_safe] += other.background.data[other.mask_safe]
+        else:
```
Comment (adonath) [inline gammapy/spectrum/dataset.py]: See comment above about the counts...

File: gammapy/spectrum/dataset.py [hunk 3]
```diff
@@ -398,6 +409,66 @@ def create(cls, e_reco, e_true=None, reference_time="2000-01-01"):
...
+        Parameters
+        ----------
+        other : `~gammapy.spectrum.SpectrumDataset`
+            the dataset to stack to the current one
+       """
+
+        if not isinstance(other, SpectrumDataset):
+            raise TypeError("Incompatible types for SpectrumDataset stacking")
+
+        if self.counts is not None:
```
Comment (adonath) [inline gammapy/spectrum/dataset.py]: I guess we should also add the check `and other.counts is not None:`?

File: gammapy/spectrum/dataset.py (hunk 1 above)
Comment (registerrier) [reply gammapy/spectrum/dataset.py]: Indeed

File: gammapy/spectrum/dataset.py (hunk 3 above)
Comment (registerrier) [reply gammapy/spectrum/dataset.py]: I wondered about this. we can indeed put `counts` to None if `other.counts is None`, but doing it silently seemed a problem to me. So I decided to let `stack` raise an error in this case. Another option would to add a `WARNING` in a log. 
Any preference?

Comment (adonath) [reply gammapy/spectrum/dataset.py]: Thinking about it again, raising an error is probably fine. So just keep as is...

File: gammapy/spectrum/dataset.py (hunk 2 above)
Comment (registerrier) [reply gammapy/spectrum/dataset.py]: Done

**Comment by registerrier** [comment]
Merging now.

**Comment by adonath** [comment]
Thanks @registerrier, I did a little bit of clean-up in https://github.com/gammapy/gammapy/commit/93b2478056368df047a081b94abd4e063b67550f.

**Pull request merged**